csharp/src/Ice/ConnectionFactory.cs

```diff
@@ -28,6 +29,10 @@ internal sealed class OutgoingConnectionFactory : IConnectionManager, IAsyncDisp
         private readonly MultiDictionary<(Endpoint, string), Connection> _connectionsByEndpoint =
             new MultiDictionary<(Endpoint, string), Connection>();
         private Task? _disposeTask;
+        // We keep a map of the connectors that has recently result in a failed connection attempt, this is used to
+        // influence the selection of connectors when creating new connections, connectors with recent failures
+        // are tried last this avoids retries choosing always connectors that have already result in a connect failure.
+        private readonly Dictionary<IConnector, DateTime> _connectFailures = new ();
```

Comment:
We could use a concurrent dictionary here instead, this would avoid to make a snapshot of the keys to remove when we purge the expired failures, we can just loop over the collection and remove expired entries.

Reply:
Fixed